src/assets/TokenDetectionController.ts

```diff
@@ -53,87 +57,157 @@ export class TokenDetectionController extends BaseController<
       ) => void;
+      onTokenListStateChange: (
+        listener: (tokenListState: TokenListState) => void,
+      ) => void;
       getBalancesInSingleCall: AssetsContractController['getBalancesInSingleCall'];
-      addTokens: TokensController['addTokens'];
+      addDetectedTokens: TokensController['addDetectedTokens'];
       getTokenListState: () => TokenListState;
       getTokensState: () => TokensState;
+      getNetworkState: () => NetworkState;
+      getPreferencesState: () => PreferencesState;
     },
     config?: Partial<TokenDetectionConfig>,
     state?: Partial<BaseState>,
   ) {
+    const {
+      provider: { chainId: defaultChainId },
+    } = getNetworkState();
+    const { useTokenDetection: defaultUseTokenDetection } =
+      getPreferencesState();
+
     super(config, state);
     this.defaultConfig = {
       interval: DEFAULT_INTERVAL,
-      networkType: MAINNET,
       selectedAddress: '',
-      tokens: [],
+      disabled: true,
+      chainId: defaultChainId,
+      isDetectionEnabledFromPreferences: defaultUseTokenDetection,
+      isDetectionEnabledForNetwork:
+        isTokenDetectionSupportedForNetwork(defaultChainId),
+      ...config,
     };
+
     this.initialize();
     this.getTokensState = getTokensState;
     this.getTokenListState = getTokenListState;
-    this.addTokens = addTokens;
-    onTokensStateChange(({ tokens }) => {
-      this.configure({ tokens });
+    this.addDetectedTokens = addDetectedTokens;
+    this.getBalancesInSingleCall = getBalancesInSingleCall;
+
+    onTokenListStateChange(({ tokenList }) => {
+      const {
+        disabled,
+        isDetectionEnabledForNetwork,
+        isDetectionEnabledFromPreferences,
+      } = this.config;
+      const hasTokens = Object.keys(tokenList).length;
+
+      if (
+        !disabled &&
+        isDetectionEnabledForNetwork &&
+        isDetectionEnabledFromPreferences &&
+        hasTokens
+      ) {
+        this.detectTokens();
+      }
     });
 
-    onPreferencesStateChange(({ selectedAddress }) => {
-      const actualSelectedAddress = this.config.selectedAddress;
-      if (selectedAddress !== actualSelectedAddress) {
-        this.configure({ selectedAddress });
+    onPreferencesStateChange(({ selectedAddress, useTokenDetection }) => {
+      const {
+        disabled,
+        selectedAddress: currentSelectedAddress,
+        isDetectionEnabledFromPreferences,
+        isDetectionEnabledForNetwork,
+      } = this.config;
+      const isSelectedAddressChanged =
+        selectedAddress !== currentSelectedAddress;
+      const isDetectionChangedFromPreferences =
+        isDetectionEnabledFromPreferences !== useTokenDetection;
+
+      this.configure({
+        isDetectionEnabledFromPreferences: useTokenDetection,
+        selectedAddress,
+      });
+
+      if (
+        !disabled &&
+        isDetectionEnabledForNetwork &&
+        useTokenDetection &&
+        (isSelectedAddressChanged || isDetectionChangedFromPreferences)
```

Comment:
`isDetectionChangedFromPreferences` is not covered in this case. I guess there is no test for the case when `useTokenDetection` changes